crypto/evp/asymcipher.c

```diff
@@ -40,55 +40,58 @@ static int evp_pkey_asym_cipher_init(EVP_PKEY_CTX *ctx, int operation,
         goto legacy;
 
     /*
-     * Ensure that the key is provided, either natively, or as a cached export.
-     *  If not, go legacy
+     * Try to derive the supported asym cipher from |ctx->keymgmt|.
      */
-    tmp_keymgmt = ctx->keymgmt;
-    provkey = evp_pkey_export_to_provider(ctx->pkey, ctx->libctx,
-                                          &tmp_keymgmt, ctx->propquery);
-    if (provkey == NULL)
-        goto legacy;
-    if (!EVP_KEYMGMT_up_ref(tmp_keymgmt)) {
+    if (!ossl_assert(ctx->pkey->keymgmt == NULL
```

Comment:
I think `pkey` could be NULL here?

Reply:
Hmmm, I haven't checked, but it seems weird if that would be NULL...  I'll look at adding some form of check for that.

Reply:
It would be an application programmer bug, but one that would be quite easy to make, e.g.

````
EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
EVP_PKEY_encrypt_init(ctx);
````

Reply:
Check added